### PR TITLE
[#110] Fixed Spec::asSingle, matchSingleResult,matchOneOrNullResult

### PR DESCRIPTION
**Pull request by adamquaile**
This is a second version of a PR for issue #110, replacing PR #111

### CODE REVIEW

File: src/EntitySpecificationRepositoryTrait.php [hunk 1]
```diff
@@ -30,6 +32,50 @@ public function match(Specification $specification, Result\ResultModifier $modif
...
+        $query = $this->getQuery($specification, $modifier);
+
+        try {
+            return $query->getSingleResult();
+        } catch (NonUniqueResultException $e) {
+            throw new Exception\NonUniqueResultException;
+        } catch (NoResultException $e) {
+            throw new Exception\NoResultException;
+        }
+    }
```
Comment (Nyholm) [inline src/EntitySpecificationRepositoryTrait.php]: Do we need our own exceptions?
We are not doing anything but re-throwing them. Shouldn't we rely on the Doctrine exceptions instead?

File: src/Result/AsSingle.php [hunk 2]
```diff
@@ -6,7 +6,7 @@
 use Doctrine\ORM\Query;
 
 /**
- * Class AsArray.
+ * Class AsSingle.
  */
 class AsSingle implements ResultModifier
```
Comment (Nyholm) [inline src/Result/AsSingle.php]: Rename this class to AsSingleScalar

File: src/Result/AsSingle.php [hunk 3]
```diff
@@ -15,6 +15,6 @@ class AsSingle implements ResultModifier
      */
     public function modify(AbstractQuery $query)
     {
-        $query->setHydrationMode(Query::HYDRATE_SINGLE_SCALAR);
+        $query->setHydrationMode(Query::HYDRATE_OBJECT);
```
Comment (Nyholm) [inline src/Result/AsSingle.php]: Keep the HYDRATE_SINGLE_SCALAR

See this comment: https://github.com/Happyr/Doctrine-Specification/pull/111#issuecomment-109919641
`Query::HYDRATE_OBJECT` is the default.

**Comment by Nyholm** [comment]
Good work! Just hade some minor questions and comments.

File: src/EntitySpecificationRepositoryTrait.php (hunk 1 above)
Comment (adamquaile) [inline src/EntitySpecificationRepositoryTrait.php]: This is to try and reduce coupling to Doctrine - I know there's a WIP PR to
that end. Happy to remove this for now if that's premature..

Comment (Nyholm) [inline src/EntitySpecificationRepositoryTrait.php]: No, okey. That is great then. Awesome :smile: 

But make sure we keep track of previous exceptions by doing something like..

```php
try {
    return $query->getSingleResult();
} catch (NonUniqueResultException $e) {
   throw new Exception\NonUniqueResultException($e->getMessage(), $e->getCode(), $e);
} catch (NoResultException $e) {
   throw new Exception\NoResultException($e->getMessage(), $e->getCode(), $e);
}
```

File: src/Result/AsSingle.php (hunk 3 above)
Comment (adamquaile) [inline src/Result/AsSingle.php]: Yes, you're right. I think I removed the wrong class.

**Comment by adamquaile** [comment]
Thanks.. I've updated both of those - probably worth giving a try in a real project just to double check.

**Comment by Nyholm** [comment]
Thank you for this. I think it's a real improvement. =)

**Comment by adamquaile** [comment]
No problem :)

**Pull request merged**